File: scala/matching-brackets/src/main/scala/MatchingBrackets.scala
object MatchingBrackets {
  def isPaired(brackets: String): Boolean = isPaired(Nil, brackets.toList)

  def isPaired(stack: List[Char], brackets: List[Char]): Boolean = brackets match {
    case (c :: cs) => (c, stack.headOption) match {
      case (']', Some('[')) | ('}', Some('{')) | (')', Some('(')) => isPaired(stack.tail, cs)
      case ('[' | '{' | '(', _) => isPaired(c :: stack, cs)
      case (']' | '}' | ')', _) => false
      case _ => isPaired(stack, cs)
    }
    case _ => stack.length == 0
  }
}

File: elixir/matching-brackets/lib/matching_brackets.ex
defmodule MatchingBrackets do
  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t()) :: boolean
  def check_brackets(str) do
    stack =
      Enum.reduce(to_charlist(str), [], fn char, stack ->
        case stack do
          nil ->
            nil

          _ ->
            [head | tail] = if stack === [], do: [nil, nil], else: stack
            case {char, head} do
              {?[, _} -> [char | stack]
              {?{, _} -> [char | stack]
              {?(, _} -> [char | stack]
              {?], ?[} -> tail
              {?}, ?{} -> tail
              {?), ?(} -> tail
              {?], _} -> nil
              {?}, _} -> nil
              {?), _} -> nil
              _ -> stack
            end
        end
      end)
    stack !== nil and stack === []
  end
end